### ISSUE TITLE
Incorrect propagatiion of type equalities in functor application?

### ISSUE THREAD

**Issue by diremy** [ocaml/ocaml#10298]
I would say that the following behavior is a bug:

    # cat id_functor_app.ml

    module type S = sig type t end
    module Res_ko =
      (functor (X : S) -> X)(struct type t = int end)

    module Res_ok2 =
      (functor (X : S) -> struct include X end) (struct type t = int end)
    module Res_ok3 =
      (functor (X : S) -> struct type t = X.t end) (struct type t = int end)

The inferred types are:

    # ocamlc -v 
    The OCaml compiler, version 4.13.0+dev0-2020-10-19
    # ocamlc -i id_functor_app.ml
    
    module type S = sig type t end
    module Res_ko : S
    module Res_ok2 : sig type t = int end
    module Res_ok3 : sig type t = int end

The type `t` is erroneously made abstract in `Res_ko`.  There is no good
reason for this, since the functor `(functor X : S -> X)` should be typed
as `(X : S) -> S/X` which is equivalent to  `(X : S) -> (sig type t = X.t end)`
and should also be the signature of the two functors used in
`Res_ok_included` and `Res_ok_included`. 

And indeed, this is the case: 

    # module M1 = (functor (X : S) -> X);;
    module M1 : functor (X : S) -> sig type t = X.t end
    # module M2 = (functor (X : S) -> struct include X end);;
    module M2 : functor (X : S) -> sig type t = X.t end
    # module M3 = (functor (X : S) -> struct type t = X.t end);;
    module M3 : functor (X : S) -> sig type t = X.t end

Besides, the type of t is concrete and therefore the avoidance problem
should not occur in this example.

**Comment by smuenzel-js** [comment]
I believe the inferred type is documented here:
https://caml.inria.fr/pub/docs/manual-ocaml/moduletypeof.html#s%3Amodule-type-of
And it looks like this is intentional: "To make this module type reusable in many situations, it is intentionally not strengthened: abstract types and datatypes are not explicitly related with the types of the original module."

**Comment by garrigue** [comment]
Actually, this entry of the manual is about the `module type of` construct, which is not used in this example.
And I don't see any formal reason here that `Res_ko` should be different from `Res_ok2` here.

Actually, this seems related to both functor and its argument having no name.
```ocaml
# module F(X : S) = X;;
module F : functor (X : S) -> sig type t = X.t end
# module Res_ok3 = F(struct type t = int end);;
module Res_ok3 : sig type t = int end
```

Looking at the code, it seems that the reason is that, when typing an application whose argument is not named, strengthening is disabled for the functor part of the application. So strengthening occurs if the functor is defined independently, but not if it is directly applied (which does not usually happen).
I can understand why strengthening is disabled when the functor is an identifier, but it is less clear why this behavior is inherited in the body of the abstraction.
Just out of curiosity, I re-enabled strengthening in the body of functors, and this fixes your problem without breaking anything in the testsuite. Maybe @xavierleroy knowns what is correct.
```diff
--- a/typing/typemod.ml
+++ b/typing/typemod.ml
@@ -2105,7 +2105,7 @@ and type_module_aux ~alias sttn funct_body anchor env smod =
           in
           Named (id, param, mty), Types.Named (id, mty.mty_type), newenv, true
       in
-      let body = type_module sttn funct_body None newenv sbody in
+      let body = type_module true funct_body None newenv sbody in
       { mod_desc = Tmod_functor(t_arg, body);
         mod_type = Mty_functor(ty_arg, body.mod_type);
         mod_env = env;
```

**Comment by xavierleroy** [comment]
With your proposed patch, if I'm not mistaken, the `sttn` parameter is always `true` and can be removed.  Then the code is back to how I wrote it in OCaml 1.00, namely all module identifiers have their types strengthened.

It's you @garrigue who introduced this special case in commit c1450ca75 as a fix for #4933.  I don't understand why not strengthening is an appropriate fix but maybe you can explain it again?

**Comment by garrigue** [comment]
Thank you @xavierleroy for reminding me that I should think of blaming myself before blaming others.
Indeed, `sttn` comes from #4933, and is an answer to the dilemma one faces when given the choice between a private abbreviation and a public one, that should subsume it, but may end up becoming abstract. Since there is only one slot to put the abbreviation, one has to choose one or the other. The public one is generally better, except if we know that it is going to be erased later through `nondep_type`.
`sttn` is precisely doing that: informing subcalls that we have no path for the argument, so that even if we strengthen the type of the functor, the types will be discarded.
In general, this is correct, however @diremy found an exception: if the functor itself is an inline lambda-abstraction which does not use its argument, the path of the argument is not used for strengthening, so there is no reason to prevent strengthening.
The patch above does avoid this by re-enabling strengthening when typing the body of functors. This does not completely remove `sttn`: it is still used, but only if the functor is a module path (which is by far the most common case).
I will submit a PR.
